extended/src/main/java/io/kubernetes/client/extended/event/legacy/EventAggregator.java

```diff
@@ -0,0 +1,107 @@
+      int maxLRUCacheEntries,
+      Function<V1Event, MutablePair<String, String>> keyFunc,
+      Function<V1Event, String> messageFunc) {
+    this.keyFunc = keyFunc;
+    this.messageFunc = messageFunc;
+    this.maxEvents = DEFAULT_MAX_EVENT_LOCAL_KEYS;
+    this.spammingCache =
+        CacheBuilder.newBuilder()
+            .maximumSize(maxLRUCacheEntries)
+            .expireAfterWrite(DEFAULT_EVENT_AGGREGATE_CACHE_EXPIRATION)
+            .build();
+  }
+
+  private Cache<String, AggregatedRecord> spammingCache;
+  private Function<V1Event, MutablePair<String, String>> keyFunc;
+  private Function<V1Event, String> messageFunc;
+
+  private final int maxEvents;
+
+  public synchronized MutablePair<V1Event, String> aggregate(V1Event event) {
+    DateTime now = DateTime.now();
+
+    MutablePair<String, String> aggregatedKeys = keyFunc.apply(event);
+    String aggregatedKey = aggregatedKeys.getLeft();
+    String localKey = aggregatedKeys.getRight();
+
+    AggregatedRecord record;
+    try {
+      record = this.spammingCache.get(aggregatedKey, AggregatedRecord::new);
+    } catch (ExecutionException e) {
+      throw new IllegalStateException(e);
+    }
+    record.lastTimestamp = now;
+    record.localKeys.add(localKey);
+
+    if (record.localKeys.size() < this.maxEvents) {
+      this.spammingCache.put(aggregatedKey, record);
+      return new MutablePair<>(event, EventUtils.getEventKey(event));
+    }
+    record.localKeys.remove(record.localKeys.stream().findAny().get()); // remove any keys
+    V1Event aggregatedEvent =
+        new V1EventBuilder(event)
+            .withMetadata(
+                new V1ObjectMetaBuilder()
+                    .withName(EventUtils.generateName(event.getInvolvedObject().getName(), now))
+                    .withNamespace(event.getInvolvedObject().getNamespace())
+                    .build())
+            .withCount(1)
+            .withFirstTimestamp(now)
+            .withLastTimestamp(now)
+            .withMessage(this.messageFunc.apply(event))
+            .build();
+    this.spammingCache.put(aggregatedKey, record);
+    return new MutablePair<>(aggregatedEvent, aggregatedKey);
+  }
+
+  private class AggregatedRecord {
+    private Set<String> localKeys = new HashSet<>();
+    private DateTime lastTimestamp;
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+
+      AggregatedRecord record = (AggregatedRecord) o;
+
+      if (localKeys != null ? !localKeys.equals(record.localKeys) : record.localKeys != null)
+        return false;
+      return lastTimestamp != null
+          ? lastTimestamp.equals(record.lastTimestamp)
+          : record.lastTimestamp == null;
+    }
+
+    @Override
+    public int hashCode() {
+      int result = localKeys != null ? localKeys.hashCode() : 0;
+      result = 31 * result + (lastTimestamp != null ? lastTimestamp.hashCode() : 0);
```

Comment:
is the 31 * result to achieve a bit shift? If so please use the bit-shift operator `<<<` for clarity.

Reply:
refactored to reuse `java.util.Objects#hashCode/equals`